lib/src/command/token_add.dart

```diff
@@ -61,4 +63,37 @@ class TokenAddCommand extends PubCommand {
       throw ApplicationException('Token is not provided within 15 minutes.');
     }
   }
+
+  Future<void> _addTokenFromStdin(Uri hostedUrl) async {
+    final token = await stdinPrompt('Enter secret token:', echoMode: false)
+        .timeout(const Duration(minutes: 15));
+    if (token.isEmpty) {
+      usageException('Token is not provided.');
+    }
+
+    tokenStore.addCredential(Credential.token(hostedUrl, token));
+    log.message(
+      'Requests to $hostedUrl will now be authenticated using the secret '
+      'token.',
+    );
+  }
+
+  Future<void> _addEnvVarToken(Uri hostedUrl) async {
+    if (envVar.isEmpty) {
+      throw DataException('Cannot use the empty string as --env-var');
+    }
```

Comment:
At a minimum we should restrict to the legal environment variable names:
   * forbidding `=` on Linux, Mac OS
   * forbidding `=` on Windows

Sources:
 * https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap08.html
 * https://docs.microsoft.com/en-us/windows/win32/procthread/environment-variables

I would suggest referencing the source in a comment explaining why we forbid `=` in the environment variable name.

------

We could further remove footguns from the user by saying that we require env vars to be:
> uppercase letters, digits, and the <underscore> ( '_' ) from the characters defined in Portable Character Set and do not begin with a digit.

Which is what the POSIX.1-2017 enforces for its variables.. But I think @sigurdm suggested that we don't artificially restrict the environment variable name. If we want to help the user we could detect if the env-var name we're given violates this rule and then warns about it.

Example:

If I explicitly pass `$MY_TOKEN` instead of `MY_TOKEN`, then maybe instead of saying:
```
dart pub token add https://some-pub.com --env-var="\$MY_TOKEN"

Requests to "https://some-pub.com" will now be authenticated using the secret 
token stored in the environment variable `$MY_TOKEN`.
Environment variable `$MY_TOKEN` is not defined.
```
We could say:

```
dart pub token add https://some-pub.com --env-var="\$MY_TOKEN"

Requests to "https://some-pub.com" will now be authenticated using the secret 
token stored in the environment variable `$MY_TOKEN`.

Environment variable `$MY_TOKEN` is not defined.

WARNING: Environment variable name given `$MY_TOKEN` starts with a dollar sign,
this is unusually for environment variable names. Are you sure you didn't want to use
the environment variable name: `MY_TOKEN`
```

We could also do:

```
dart pub token add https://some-pub.com --env-var="MY_TOKEN "

Requests to "https://some-pub.com" will now be authenticated using the secret 
token stored in the environment variable `MY_TOKEN `.

Environment variable `MY_TOKEN ` is not defined.

WARNING: Environment variable name given `MY_TOKEN ` starts with a dollar sign,
this is unusually for environment variable names. Are you sure you didn't want to use
the environment variable name: `MY_TOKEN `
```
and maybe even:
```
export MY_TOKEN="bestpassword1234"
dart pub token add https://some-pub.com --env-var="$MY_TOKEN"  # <-- I accidentally gave the value of the env var instead of the name

Requests to "https://some-pub.com" will now be authenticated using the secret 
token stored in the environment variable `bestpassword1234`.

Environment variable `bestpassword1234` is not defined.

WARNING: Environment variable name given `bestpassword1234` does not use
upper-case characters, 0-9, and underscore. If you didn't want to use this environment variable name
you can overwrite it with:
  dart pub token add https://some-pub.com --env-var=VARIABLE_NAME
Or remove it using:
  dart pub token remove https://some-pub.com
```

Warnings like this wouldn't be perfect, but they would certainly help cases where people accidentally pass the value of an environment variable rather than the actual env var name. Or pass the name, but add extract spaces or prefix it with a dollar sign by accident. These things are easy to accidentally do, and unless you read carefully, they can be hard to debug.